Initial workload identity section (#36)
ietf-wg-wimse/draft-ietf-wimse-arch: draft-ietf-wimse-arch.md

## Workload Identity

> A workload identity may be composed of many attributes describing different aspects of a workload. These attributes may describe different sets of information such as a workload's role within a system, the software that the workload runs or the hardware that the workload runs on. Sets of attributes may be determined by different authorities running in different parts of the system. WIMSE defines a **Workload Identifier** to used in representing and referencing the workload.  It allows different authorities of sets identity attributes to reference the workload in a consistent way.

Reviewer: Typo in "to used in representing". Maybe "to be used in representing"?
Typo in "different authorities of sets"... Maybe "different authorities to set"? The full sentence may revise.

> A workload identity may be composed of many attributes describing different aspects of a workload. These attributes may describe different sets of information such as a workload's role within a system, the software that the workload runs or the hardware that the workload runs on. Sets of attributes may be determined by different authorities running in different parts of the system. WIMSE defines a **Workload Identifier** to used in representing and referencing the workload.  It allows different authorities of sets identity attributes to reference the workload in a consistent way.

Reviewer 2: ```suggestion
Workload identity often comprises multiple attributes that describe various aspects of a workload. These attributes can encompass diverse sets of information, including the workload's role within a system, the software it operates, and the hardware environment it utilizes. Different authorities across various parts of the system define these attribute sets. This architecture introduces a Workload Identifier for representing and referencing workloads, enabling diverse authorities with distinct identity attribute sets to consistently reference the workload.
```

> The Workload Identifier consists of a concise string allocated within a namespace defined by a Trust Domain. This Workload Identifier is present in Workload Identity Tokens and X.509 certificates issued by the authority for the Trust Domain. The Workload Identifier is used to associate additional identity attributes to the workload through the use of tokens (workload attribute tokens?) or online look up services. It may also be used directly in authorization calculations and audit logs.

Reviewer 3: >  The Workload Identifier is used to associate additional identity attributes to the workload through the use of tokens (workload attribute tokens?) or online look up services.

Should this be the workload credential (X.509 cert or JWT)? Although the identifier can include attribute information, I would typically expect to find other claims and properties about the workload in the credential they represent (especially if those claims may change or evolve over time - like version or bill of materials).

> The Workload Identifier consists of a concise string allocated within a namespace defined by a Trust Domain. This Workload Identifier is present in Workload Identity Tokens and X.509 certificates issued by the authority for the Trust Domain. The Workload Identifier is used to associate additional identity attributes to the workload through the use of tokens (workload attribute tokens?) or online look up services. It may also be used directly in authorization calculations and audit logs.

Reviewer 2: ```suggestion
The Workload Identifier is a concise string allocated within a namespace defined by a Trust Domain. It appears in Workload Identity Tokens and X.509 certificates issued by a Trust Domain. This identifier facilitates the association of additional identity attributes with the workload through tokens (such as workload attribute tokens) or online lookup services. Moreover, it plays a role in authorization decisions and for use in audit logs.
```

> The Trust Domain consists of a string that matches the format of a fully qualified domain name. It is the intent that a Trust Domain is actually a domain name registered to the organization defining the Trust Domain, but this may not be true in all cases. The Trust Domain also maps to the issuer of cryptographically signed Workload Identity Tokens (WIT) or X.509 Certificate. The association between a Trust Domain and the cryptographic root of the signing authority for that Trust Domain must be made securely through an out-of-band mechanisms. [TODO: where should mechanisms be defined?]

Author: I left it open here, but I really think the trust domain should be represented within the identifier in a consistent way, however alternatively it could be a separate identifier for a trust domain that mist be represented in a certificate or token field.

> The Trust Domain consists of a string that matches the format of a fully qualified domain name. It is the intent that a Trust Domain is actually a domain name registered to the organization defining the Trust Domain, but this may not be true in all cases. The Trust Domain also maps to the issuer of cryptographically signed Workload Identity Tokens (WIT) or X.509 Certificate. The association between a Trust Domain and the cryptographic root of the signing authority for that Trust Domain must be made securely through an out-of-band mechanisms. [TODO: where should mechanisms be defined?]

Reviewer 2: ```suggestion
The Trust Domain is represented as a string formatted like a fully qualified domain name (FQDN). While ideally linked to an organization's registered domain name, this may vary. The Trust Domain maps to the issuer of cryptographically signed Workload Identity Tokens (WITs) or X.509 certificates. The secure association between a Trust Domain and the cryptographic root of the signing authority must be established using out-of-band mechanisms (TODO: Specify where these mechanisms should be defined).
```

> The Trust Domain also defines how the rest of the Workload Identifier is constructed. The Workload Identifier may represent a type of workload such that the same identifier may be used by many instances of the same service. A Trust Domain may choose identifiers to represent a specific instance of a workload such that each workload of the same type will have a specific identity.  The Trust Domain could choose a naming scheme that allows for both objects by imposing a hierarchical structure on the naming format.

Reviewer 3: Perhaps consider splitting out the definition of a trust domain (what it does) from the formatting of how a trust domain may be represented (how). Are there any differences in the trust domain format here that your thinking of compared to a SPIFFE trust domain?

> The Trust Domain also defines how the rest of the Workload Identifier is constructed. The Workload Identifier may represent a type of workload such that the same identifier may be used by many instances of the same service. A Trust Domain may choose identifiers to represent a specific instance of a workload such that each workload of the same type will have a specific identity.  The Trust Domain could choose a naming scheme that allows for both objects by imposing a hierarchical structure on the naming format.

Author: @PieterKas good suggestion, I will work on this. 

@arndt-s  My thinking: The issuing of identities must be coordinated so there are no conflicts.  Therefore it seems that it is best if the format is determined within the trust domain.  Issuers of identity credentials within the trust domain will generate identifiers for workloads, there may be more than one issuer so the format of the identifier.  The overall scheme is defined by the trust domain, portions of the name space can be delegated to specific issuers.
perhaps this is what you mean?

> The Trust Domain also defines which mechanisms are used to initially bootstrap a workload with a Workload Identifier and the mechanisms for a workload to obtain its workload identity credentials in the form of X.509 certificates and Workload Identity Tokens.

Reviewer 2: ```suggestion
Additionally, the Trust Domain defines the mechanisms for initially bootstrapping a workload with a Workload Identifier and the procedures for obtaining workload identity credentials in the form of X.509 certificates and Workload Identity Tokens.
```
